### PR TITLE
fixes GH-166:

### PR DESCRIPTION
**Pull request by shekyan**
- handle `,` and `;` characters in source expression
- refactor path matching algorithm to reflect latest specification

### CODE REVIEW

**Comment by michaelficarra** [comment]
Needs a test for constructing a `HostSource` with a `,` or `;` in it and adding it to an existing policy. The policy should be safely escaped when rendered.

We should also make sure that path matches are escape-agnostic. What I meant by that is that unnecessarily-escaped characters in the path match their unescaped equivalents. It looks like you've implemented [§6.6.1.10](https://w3c.github.io/webappsec-csp/#match-paths) pretty directly, so it should already behave this way; we just need tests.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java [hunk 1]
```diff
@@ -128,4 +129,58 @@ public boolean matchesOnlyOrigin(@Nonnull SchemeHostPortTriple origin) {
...
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static String decodeString(@Nonnull String s) {
+        String out = "";
+        try {
+            out = URLDecoder.decode(s,"UTF-8");
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: Space after comma.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java [hunk 2]
```diff
@@ -128,4 +129,58 @@ public boolean matchesOnlyOrigin(@Nonnull SchemeHostPortTriple origin) {
...
+            String b = decodeString((String) it2.next());
+            if (!a.equals(b)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static String decodeString(@Nonnull String s) {
+        String out = "";
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: I might implement this function like this:

```java
public static String decodeString(@Nonnull String s) {
    try {
        return URLDecoder.decode(s, "UTF-8");
    } catch (Exception e) {
        return s;
    }
}
```

The reason is that illegal sequences such as `%%` or `%0G` would probably cause this to throw, but should be preserved as-is instead. We should have tests. My concern with the implementation I've suggested is that a path part containing both illegal sequences and legal sequences should still have the replacement done on the legal sequences. We may have to come up with a custom solution.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java [hunk 3]
```diff
@@ -128,4 +129,58 @@ public boolean matchesOnlyOrigin(@Nonnull SchemeHostPortTriple origin) {
...
+        try {
+            out = URLDecoder.decode(s,"UTF-8");
+        } catch (Exception e) {}
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: Don't catch `Exception`. Catch whatever it is that's thrown by this function.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java [hunk 4]
```diff
@@ -128,4 +129,58 @@ public boolean matchesOnlyOrigin(@Nonnull SchemeHostPortTriple origin) {
...
+        if (pathA.matches("/") && pathB.isEmpty()) {
+            return true;
+        }
+
+        boolean exactMatch = pathA.endsWith("/") ? false : true;
+
+        List<String> pathListA = Arrays.asList(pathA.split("/"));
+        List<String> pathListB = Arrays.asList(pathB.split("/"));
+
+        long countA = pathA.chars().filter(ch -> ch =='/').count();
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: I don't understand the reason for the `countA` and `countB` variables and their usages. It seems that if they were removed you'd be following §6.6.1.10 directly.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java [hunk 5]
```diff
@@ -9,7 +9,9 @@
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
-import java.util.Objects;
+import java.net.URLDecoder;
+import java.util.*;
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: Don't import `*` unless there's over 100 imports from that namespace.

Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: IntelliJ magic...

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java (hunk 4 above)
Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: The reason is https://github.com/w3c/webappsec-csp/issues/171 ;)
I'll make a patch to the spec once back from the snow..

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java (hunk 2 above)
Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: Agreed.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java (hunk 4 above)
Comment (shekyan) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: @michaelficarra , welcome to improve the count logic. Didn't have much time for more elegant solution, but the idea is that strict splitting loses info on delimiters, and just counting them should be enough

**Comment by shekyan** [comment]
> Needs a test for constructing a HostSource with a , or ; in it and adding it to an existing policy. The policy should be safely escaped when rendered.
I think https://github.com/shapesecurity/salvation/pull/167/files#diff-2bb2d7b2fa7ffdef22812c8d231f7337R1072 is minimally covers it but sure, will add explicit tests for that.

**Comment by shekyan** [comment]
Addressed comments

File: src/test/java/com/shapesecurity/salvation/ParserTest.java [hunk 6]
```diff
@@ -209,6 +209,16 @@
...
         p = parse("script-src 'nonce-1234'; style-src 'nonce-1234'; default-src a");
         assertEquals("script-src 'nonce-1234'; style-src 'nonce-1234'; default-src a", p.show());
+
+        p = parse("script-src http://example.org/cpproducts/index.htm;jsessionid=0C159CF2854B7519C5379634FAC36577");
+        assertEquals("script-src http://example.org/cpproducts/index.htm", p.show());
+
+        p = parse("script-src http://example.org/cpproducts/index.htm%3Bjsessionid=0C159CF2854B7519C5379634FAC36577");
+        assertEquals("script-src http://example.org/cpproducts/index.htm%3Bjsessionid=0C159CF2854B7519C5379634FAC36577", p.show());
+
+        p = parse("script-src http://example.org/cpproducts/index.htm%3Bjsessionid=0C159CF2854B7519C5379634FAC36577");
```
Comment (michaelficarra) [inline src/test/java/com/shapesecurity/salvation/ParserTest.java]: Is this the same test as the one above it?

**Comment by michaelficarra** [comment]
We still need tests for invalid escapes in the paths. And tests for mixing them with valid escapes.

File: src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java (hunk 4 above)
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/salvation/directiveValues/HostSource.java]: Let's hold off on this PR until that issue is resolved.

**Pull request merged**